---
title: "CG and CH contexts"
output: html_document
params:
  seed: 11
fig_caption: yes
---
    
```{r settings, include=FALSE}

library(knitr)
## library(devtools)
library(reshape2)
## library(lattice)
## library(pheatmap)
library(seqLogo)
## library(margins)
## library(data.table)
library(ggplot2)
library(ggfortify)
library(ggseqlogo)
library(gridExtra)

TASK <- "cg_context"
HOME <- '/home/imallona'
WD <-  file.path(HOME, 'mnt', 'nfs', TASK)
DATA <- file.path(HOME, 'data')

MIN_DEPTH <- 10
NFS <- file.path(HOME, 'mnt', 'nfs')

beta2m <- function(beta) {
    m <- log2(beta/(1 - beta))
    return(m)
}

m2beta <- function(m) {
    beta <- 2^m/(2^m + 1)
    return(beta)
}

opts_chunk$set(fig.width = 12,
               fig.height = 12,
               fig.path = file.path(WD, TASK, 'RmdFigs'),
               output.dir = WD,
               root.dir = WD,
               cache = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               message = TRUE)

dir.create(file.path(WD, TASK))

```


```{r design, include = TRUE, cache = TRUE, fig.width = 20}
knitr::include_graphics("/home/imallona/src/cg_context/stranded_dnameth.png")
```


```{r run_trace}
sources <- list.files(file.path(WD), recursive = TRUE, pattern = '*sources')

for (fn in sources) {
    cat(sprintf('SOURCE BAMS FOR CONFIG FILE %s', fn))
    cat('\n')
    for (line in readLines(file.path(WD, fn))) {
        cat(sprintf('%s\n', line))
    }
    cat('\n')
    cat('\n')
}

```

```{r config_annot, include = FALSE}

samples_annot <- read.table(text ='sample,genotype,seq
20151223.B-MmES_TKOD3A1c1-3_R,tko+d3a1,bwa_hiseq2k
BSSE_QGF_16209_131212_SN792_0303_AD2AJ9ACXX_6_ACAGTGA_L006_001,tko+d3a2,bwa_hiseq2k
BSSE_QGF_16209_131212_SN792_0303_AD2AJ9ACXX_lane6_Undetermined_L006_001,tko+d3a2,bwa_hiseq2k
BSSE_QGF_16209_131212_SN792_0303_AD2AJ9ACXX_lane6_Undetermined_L006_002,tko+d3a2,bwa_hiseq2k
BSSE_QGF_16209_131212_SN792_0303_AD2AJ9ACXX_lane6_Undetermined_L006_003,tko+d3a2,bwa_hiseq2k
SRR1274743,tko+d3a2,bwa_miseq_pe
BSSE_QGF_16210_131212_SN792_0303_AD2AJ9ACXX_7_CAGATCA_L007_001,tko+d3b1,bwa_hiseq2k
BSSE_QGF_16210_131212_SN792_0303_AD2AJ9ACXX_lane7_Undetermined_L007_001,tko+d3b1,bwa_hiseq2k
BSSE_QGF_16210_131212_SN792_0303_AD2AJ9ACXX_lane7_Undetermined_L007_002,tko+d3b1,bwa_hiseq2k
BSSE_QGF_16210_131212_SN792_0303_AD2AJ9ACXX_lane7_Undetermined_L007_003,tko+d3b1,bwa_hiseq2k
SRR1274744,tko+d3b1,bwa_miseq_pe
SRR1274745,tko+d3b1,bwa_miseq_pe
SRR1653150,qko+d3b1,bwa_hiseq2k
SRR1653151,qko+d3b1,bwa_hiseq2k
SRR1653152,qko+d3b1,bwa_hiseq2k
SRR1653153,qko+d3b1,bwa_hiseq2k
SRR1653154,qko+d3b1,bwa_hiseq2k
SRR1653155,qko+d3b1,bwa_hiseq2k
SRR1653156,qko+d3b1,bwa_hiseq2k
SRR1653157,qko+d3b1,bwa_hiseq2k
SRR1653158,qko+d3b1,bwa_hiseq2k
SRR1653159,qko+d3b1,bwa_hiseq2k
SRR1653160,qko+d3b1,bwa_hiseq2k
SRR1653161,qko+d3b1,bwa_hiseq2k
SRR1653162,qko+d3b1,bwa_miseq_pe
', header = TRUE, sep = ',')

```

```{r data_import, include = TRUE, echo = FALSE, cache = TRUE, cache.lazy = FALSE}

## meth and unmeth separately
d <- list(cg = list(),
          ch = list())

for (context in c('cg', 'ch')) {
    for (genotype in dirname(sources)) {
        for (state in c('meth', 'unmeth')) {

            fd <- read.table(
                file.path(WD, genotype,
                          sprintf('motif_counts_%s_%s.txt', context, state)),
                col.names = c('count', 'motif'),
                stringsAsFactors = FALSE)

            ## removing Ns and short motifs
            rownames(fd) <- fd$motif
            fd <- fd[grep('N', fd$motif, invert = TRUE),]
            fd <- fd[nchar(fd$motif) == 8,]
            fd <- fd[sort(fd$motif),]
            
            d[[context]][[genotype]][[state]] <- fd
            rm(fd)
        }
    }
}

## dictionary of the whole set of motifs
nucl <- c('T', 'C', 'G', 'A')

mdict <- list('cg' = NULL, 'ch' = NULL)
mdict$cg <- apply(expand.grid('1' = nucl, '2' = nucl, '3' = nucl,
                               '4' = 'C',  '5' = 'G',
                               '6' = nucl, '7' = nucl, '8' = nucl ),
                   1,
                   function(x) paste(x, collapse = ''))



mdict$ch <- apply(expand.grid('1' = nucl, '2' = nucl, '3' = nucl,
                               '4' = 'C',  '5' = c('A', 'T', 'C'),
                               '6' = nucl, '7' = nucl, '8' = nucl),
                   1,
                   function(x) paste(x, collapse = ''))

for (item in names(mdict)) {
    fd <- mdict[[item]]
    fd <- sort(as.character(fd))
    fd <- data.frame(motif = fd, count = 0)
    rownames(fd) <- fd$motif
    mdict[[item]] <- fd
    rm(fd)
}

## filling with zeroes the motifs that are not present in some datasets


for (context in c('cg', 'ch')) {
    for (genotype in dirname(sources)) {
        for (state in c('meth', 'unmeth')) {
            curr <- d[[context]][[genotype]][[state]]

            mis <- setdiff(rownames(mdict[[context]]), rownames(curr))
            curr <- rbind(curr, mdict[[context]][mis,])
            ## lexicographical sorting
            curr <- curr[order(as.character(curr$motif)),]
            d[[context]][[genotype]][[state]] <- curr
            
            
            stopifnot(nrow(curr) == nrow(mdict[[context]]))
            rm(curr, mis)
        }
    }
}

```

```{r normalize}

## normalization factor based upon the number of meth Cs vs the number of unmeth Cs
nf <- data.frame(genotype = rep(dirname(sources), each = 2),
                   context = c('cg', 'ch'),
                   meth = NA,
                   unmeth = NA,
                   ratio = NA)
                   
for (i in 1:nrow(nf)) {
    for (status in c('meth', 'unmeth')) {
        nf[i, status] <- sum(d[[nf[i, 'context']]][[nf[i, 'genotype']]][[status]]$count)
    }
    nf[i, 'ratio'] <- nf[i, 'meth']/(nf[i, 'meth'] + nf[i, 'unmeth'])
}

## apply the normalization factor


nd <- list(cg = setNames(data.frame(matrix(nrow = length(mdict$cg$motif),
                                           ncol = length(sources) + 1)),
                         c('motif', dirname(sources))),
           ch = setNames(data.frame(matrix(nrow = length(mdict$ch$motif),
                                           ncol = length(sources) + 1)),
                         c('motif', dirname(sources))))

for (context in names(nd)) {
    nd[[context]][,'motif'] <-  as.character(mdict[[context]]$motif)
}
           

for (context in c('cg', 'ch')) {
    for (genotype in dirname(sources)) {
                
        stopifnot(d[[context]][[genotype]][['meth']]$motif ==
                  d[[context]][[genotype]][['muneth']]$motif)
        
        unnormalized <- d[[context]][[genotype]][['meth']]$count /
            d[[context]][[genotype]][['unmeth']]$count


        normalized <- unnormalized / nf[nf$genotype == genotype
                                        & nf$context == context, 'ratio']

        stopifnot(d[[context]][[genotype]][['meth']]$motif ==
                  nd[[context]]$motif)
        
        nd[[context]][,genotype] <- normalized
            
    }

}

## is this ok?
apply(nd[[1]][,2:5], 2, mean)
## merged_qko+d3b1 merged_tko+d3a1 merged_tko+d3a2 merged_tko+d3b1 
##        1.487654        2.050861        1.567780        1.186425 

## compare with the old integration outputs

## plot

for (context in names(nd)) {
    melted <- melt(nd[[context]])
    save(melted, file = sprintf('test_plots%s.RData', context))

    ## sorted by kmer

    print(ggplot(data = melted, aes(x = motif, y = value, group = variable, color = variable))+ 
        geom_point())

    ## sorted by average dnameth by kmer
    ## melted <- melted[ order(melted$value),]


    ## sorting them from low to high meth
    ## sorted <- names(sort(with(melted, tapply(value, motif, mean))))

    ## mtcars$cyl3 <- with(mtcars, reorder(cyl, cyl, function(x) -length(x)))
    ## ggplot(mtcars, aes(cyl3)) + geom_bar()

    melted$fac = reorder(melted$motif, melted$value, mean)

    ## ggplot(d, aes(x=reorder(fac, y, mean), y=y))
    print(ggplot(data = melted, aes(reorder(fac, value, mean), y = value, group = variable,
                              color = variable))+ 
        geom_point()  )
    
}


```

```{r differential}


## picking the top 10 motifs with the most variability

## the most variable motifs




topmotifs <- list()
for (context in names(nd)) {

    vars <- data.frame(motif = nd[[context]]$motif,
                       var = apply(nd[[context]][2:ncol(nd[[context]])], 1, var))
    vars <- vars[order(vars$var, decreasing = TRUE),]
    plot(density(vars$var))

    ## top 1% variable motifs
    top <- vars[vars$var >= quantile(vars$var, probs = 0.99),]
    selected <-  nd[[context]][nd[[context]]$motif %in% top$motif,]
    rownames(selected) <- selected$motif
    selected <- selected[,-1]

    topmotifs[[context]][['q99']] <- as.character(top$motif)
    
    heatmap(as.matrix(selected),
            Rowv = as.dendrogram(hclust(dist(as.matrix(selected)), method = 'ward')),
            Colv = as.dendrogram(hclust(dist(t(as.matrix(selected))), method = 'ward')),
            main = sprintf('q99 variable motifs, n = %s', nrow(selected)))

    ## now pca
    print(autoplot(prcomp(selected), label = TRUE,
             main = sprintf('q99 variable motifs, n = %s', nrow(selected))))
    print(autoplot(prcomp(t(selected)), label = TRUE, loadings = TRUE,
             loadings.label = TRUE, loadings.label.size = 3,
             main =  sprintf('q99 variable motifs, n = %s', nrow(selected))))

    ## pca on the whole stuff
    rownames(nd[[context]]) <- nd[[context]]$motif
    print(autoplot(prcomp(nd[[context]][,-1]), label = TRUE,
             main = 'all motifs'))
    
    print(autoplot(prcomp(t(nd[[context]][,-1])), label = TRUE, loadings = TRUE,
             loadings.label = TRUE, loadings.label.size = 3,
             main =  'all motifs'))


    for (topn in c(10, 20, 50)) {
        top <- head(as.character(vars$motif), topn)

        selected <- nd[[context]][nd[[context]]$motif %in% top,]
        rownames(selected) <- selected$motif
        selected <- selected[,-1]
        
        heatmap(as.matrix(selected),
                Rowv = as.dendrogram(hclust(dist(as.matrix(selected)), method = 'ward')),
                Colv = as.dendrogram(hclust(dist(t(as.matrix(selected))), method = 'ward')),
                main = sprintf('top %s variable motifs', n = topn))

        topmotifs[[context]][[as.character(topn)]] <- top

        ## m1 <- lm(value ~ motif * variable, data = selected,])
    } 
}

## the motifs that show the highest difference in a genotype

```

```{r table_plot_tops_cg}
for (context in names(topmotifs)) {
    for (cutoff in names(topmotifs[[context]])) {
          selected <- nd[[context]][nd[[context]]$motif %in% topmotifs[[context]][[cutoff]],]  
    }


    ## sorted lexicographically
    
    melted <- melt(selected)
    print(ggplot(data = melted, aes(x = motif, y = value, group = variable, color = variable))+ 
        geom_point()+
        theme(axis.text.x = element_text(angle = 90, hjust = 1)))

    ## sorted by trend

    melted$fac = reorder(melted$motif, melted$value, mean)

    ## ggplot(d, aes(x=reorder(fac, y, mean), y=y))
    print(ggplot(data = melted, aes(reorder(fac, value, mean), y = value, group = variable,
                              color = variable))+ 
        geom_point()+
        theme(axis.text.x = element_text(angle = 90, hjust = 1)))

}

```


```{r logos}

for (context in names(topmotifs)) {
    for (cutoff in names(topmotifs[[context]])) {
        selected <- nd[[context]][nd[[context]]$motif %in% topmotifs[[context]][[cutoff]],]
        melted <- melt(selected)

        ## setting the item with min value as 1 and scoring others (truncating)
        melted$score <- round(melted$value/min(melted$value))
        lprobs <- list()
        for (genotype in levels(melted$variable)) {
            curr <- melted[melted$variable == genotype,]
            
            
            lprobs[[genotype]] <- rep(curr$motif, curr$score)
        
            
        }

        print(sprintf('context %s cutoff top %s', context, cutoff))
        print(ggseqlogo(lprobs, ncol = 1, method = 'probability'))
        print(ggseqlogo(lprobs, ncol = 1, method = 'bits'))
        rm(selected, lprobs)

    }
}

## gridExtra::grid.arrange(lbits[[1]], lbits[[2]], lbits[[3]], lbits[[4]], ncol = 1)

## ggseqlogo(lbits, ncol=1)

```


```{r sessionInfo}
date()
sessionInfo()
devtools::session_info()

```
