---
title: "CG and CH contexts w/ quickchanges"
author: "Izaskun Mallona (Tuncay Baubec lab)"
output:
  html_document:
    keep_md: true
    toc: true
    toc_float: true
    toc_depth: 4
params:
  seed: 11

---
    
```{r settings, include=FALSE}

library(knitr)
## library(devtools)
library(reshape2)
## library(lattice)
library(pheatmap)
library(seqLogo)
## library(margins)
## library(data.table)
library(ggplot2)
library(ggfortify)
library(ggseqlogo)
library(gridExtra)

TASK <- "cg_context"
HOME <- '/home/imallona'
WD <-  file.path(HOME, 'mnt', 'nfs', TASK)
DATA <- file.path(HOME, 'data')

MIN_DEPTH <- 10
NFS <- file.path(HOME, 'mnt', 'nfs')

beta2m <- function(beta) {
    m <- log2(beta/(1 - beta))
    return(m)
}

m2beta <- function(m) {
    beta <- 2^m/(2^m + 1)
    return(beta)
}

opts_chunk$set(fig.width = 12,
               fig.height = 12,
               fig.path = file.path(WD, TASK, 'RmdFigs'),
               output.dir = WD,
               root.dir = WD,
               cache = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               message = TRUE)

dir.create(file.path(WD, TASK))

```

# Aim

- Hypothesis: DNMTs (DNA meth transferases) show sequence specificity
-   Datasets: WGBS in three mouse cells lacking DNA meth (called
    `triple KOs or TKOs`)
-   They have a different DNMT (DNA meth transferases) addback each

# Design

```{r design, include = TRUE, cache = TRUE, fig.width = 20}
knitr::include_graphics("/home/imallona/src/cg_context/stranded_dnameth.png")

```

-   Flow
    -   Call DNA methylation for every CpG in a strand-specific manner
    -   Collapse counts (methylated/unmethylated) for each context 6-mer
    `NNNCGNNN`
    - Get nucleotide proportions for each position (i.e. 1 to 8) for methylated and unmethylated motifs separately
    - Get a log2 ratio

  
# Datasets load

```{r run_trace}
sources <- list.files(file.path(WD), recursive = TRUE, pattern = '*sources')

for (fn in sources) {
    cat(sprintf('SOURCE BAMS FOR CONFIG FILE %s', fn))
    cat('\n')
    for (line in readLines(file.path(WD, fn))) {
        cat(sprintf('%s\n', line))
    }
    cat('\n')
    cat('\n')
}

```


```{r data_import, include = TRUE, echo = FALSE, cache = TRUE, cache.lazy = FALSE}

## meth and unmeth separately
d <- list(cg = list(),
          ch = list())

for (context in c('cg', 'ch')) {
    for (genotype in dirname(sources)) {
        for (state in c('meth', 'unmeth')) {

            fd <- read.table(
                file.path(WD, genotype,
                          sprintf('motif_counts_%s_%s.txt', context, state)),
                col.names = c('count', 'motif'),
                stringsAsFactors = FALSE)

            ## removing Ns and short motifs
            rownames(fd) <- fd$motif
            fd <- fd[grep('N', fd$motif, invert = TRUE),]
            fd <- fd[nchar(fd$motif) == 8,]
            fd <- fd[sort(fd$motif),]
            
            d[[context]][[genotype]][[state]] <- fd
            rm(fd)
        }
    }
}

```


## Quickchanges QC 

Quickchanges are named WGBS 3 and 4, no idea which are the aminoacid changes and which was the background.

Regarding sequence cutadapting/sickle-ing

```
[imallona@imlstaupo quickchanges]$ pwd
/home/imallona/mnt/nfs/cg_context/quickchanges

[imallona@imlstaupo quickchanges]$ find . -name "*sickle.log" | xargs head
==> ./20180913.A-WGBS_3_cutadapt_sickle.log <==

FastQ paired records kept: 109032024 (54516012 pairs)
FastQ single records kept: 479178 (from PE1: 455320, from PE2: 23858)
FastQ paired records discarded: 17340 (8670 pairs)
FastQ single records discarded: 479178 (from PE1: 23858, from PE2: 455320)


==> ./20180913.A-WGBS_4_cutadapt_sickle.log <==

FastQ paired records kept: 152820834 (76410417 pairs)
FastQ single records kept: 609665 (from PE1: 585466, from PE2: 24199)
FastQ paired records discarded: 25058 (12529 pairs)
FastQ single records discarded: 609665 (from PE1: 24199, from PE2: 585466)

```

As for mapping (bwameth default) after dedup and map40 filtering

```
[imallona@imlstaupo quickchanges]$ samtools flagstat -@ 12 20180913.A-WGBS_3_bwameth_default.bam
57261802 + 4723556 in total (QC-passed reads + QC-failed reads)
4011243 + 215900 secondary
0 + 0 supplementary
0 + 0 duplicates
57261802 + 4723556 mapped (100.00% : 100.00%)
53250559 + 4507656 paired in sequencing
26854450 + 2289387 read1
26396109 + 2218269 read2
52410363 + 0 properly paired (98.42% : 0.00%)
53151977 + 4466585 with itself and mate mapped
98582 + 41071 singletons (0.19% : 0.91%)
184465 + 118036 with mate mapped to a different chr
184465 + 118036 with mate mapped to a different chr (mapQ>=5)


[imallona@imlstaupo quickchanges]$ samtools flagstat -@ 12 20180913.A-WGBS_4_bwameth_default.bam
59395798 + 3890657 in total (QC-passed reads + QC-failed reads)
2950326 + 202473 secondary
0 + 0 supplementary
0 + 0 duplicates
59395798 + 3890657 mapped (100.00% : 100.00%)
56445472 + 3688184 paired in sequencing
28582565 + 1895496 read1
27862907 + 1792688 read2
55602134 + 0 properly paired (98.51% : 0.00%)
56371508 + 3664469 with itself and mate mapped
73964 + 23715 singletons (0.13% : 0.64%)
383760 + 258508 with mate mapped to a different chr
383760 + 258508 with mate mapped to a different chr (mapQ>=5)

```

So around 60M alignments with highest quality. (BWA sets as QC-failed when reads are soft-clipped)

Secondary alignments are half of them (discarded with methyldackel)

Regarding duplication

```
[imallona@imlstaupo quickchanges]$ find . -name "*metrics" | xargs fgrep LIBRARY -A 1


./20180913.A-WGBS_3_bwameth_default_dup_marked.metrics:LIBRARY	UNPAIRED_READS_EXAMINED	READ_PAIRS_EXAMINED	SECONDARY_OR_SUPPLEMENTARY_RDS	UNMAPPED_READS	UNPAIRED_READ_DUPLICATES	READ_PAIR_DUPLICATES	READ_PAIR_OPTICAL_DUPLICATES	PERCENT_DUPLICATION	ESTIMATED_LIBRARY_SIZE
./20180913.A-WGBS_3_bwameth_default_dup_marked.metrics-Unknown Library	290264	54087918	14857545	565924	103908	14880118	141779   0.275332	79780960


./20180913.A-WGBS_4_bwameth_default_dup_marked.metrics:LIBRARY	UNPAIRED_READS_EXAMINED	READ_PAIRS_EXAMINED	SECONDARY_OR_SUPPLEMENTARY_RDS	UNMAPPED_READS	UNPAIRED_READ_DUPLICATES	READ_PAIR_DUPLICATES	READ_PAIR_OPTICAL_DUPLICATES	PERCENT_DUPLICATION	ESTIMATED_LIBRARY_SIZE
./20180913.A-WGBS_4_bwameth_default_dup_marked.metrics-Unknown Library	282352	76064994	13934541	408494	147585	32994796	208542   0.433936	60046770

```

So duplication rates of 27 to 43%.

Qualimap run as well before deduplication but after mapq subsetting value 40. Oddly enough it reports back mapq values of 30 etc.

```
cd ~/cg_context
mkdir qualimap_quickchanges
cd $_

rsync -avt imlstaupo://home/imallona/mnt/nfs/cg_context/quickchanges/*qualimap .

```

Checking mapqvalues distribution with samtools

```

samtools view -@ 12 -q 10 -b 20180913.A-WGBS_3_bwameth_default.bam | wc -l
# 17131607
samtools view -@ 12 -q 30 -b 20180913.A-WGBS_3_bwameth_default.bam | wc -l
# 17131607
```

Is bamqc broken then? But results are consistent with the `cphg_quickchanges.sh` protocol, so nothing to see here.

```{r data_import_quickchanges}
## these are located at the quickchanges folder. No idea which one is it, names are 3 and 4.
for (context in c('cg', 'ch')) {
    for (genotype in c('20180913.A-WGBS_3', '20180913.A-WGBS_4')) {
        for (state in c('meth', 'unmeth')) {
            fd <- read.table(
                file.path(WD, 'quickchanges',
                          sprintf('%s_motif_counts_%s_%s_%s.txt', genotype, context, state,
                                  genotype)),
                col.names = c('count', 'motif'),
                stringsAsFactors = FALSE)

            ## removing Ns and short motifs
            rownames(fd) <- fd$motif
            fd <- fd[grep('N', fd$motif, invert = TRUE),]
            fd <- fd[nchar(fd$motif) == 8,]
            fd <- fd[sort(fd$motif),]
            
            d[[context]][[genotype]][[state]] <- fd
            rm(fd)
        }
    }
}

```


```{r updating_sources_quickchanges}

sources <- c(sources, '20180913.A-WGBS_3/20180913.A-WGBS_3', '20180913.A-WGBS_4/20180913.A-WGBS_4')

```

``` {r preproc_countings}
## dictionary of the whole set of motifs
nucl <- c('T', 'C', 'G', 'A')

mdict <- list('cg' = NULL, 'ch' = NULL)
mdict$cg <- apply(expand.grid('1' = nucl, '2' = nucl, '3' = nucl,
                               '4' = 'C',  '5' = 'G',
                               '6' = nucl, '7' = nucl, '8' = nucl ),
                   1,
                   function(x) paste(x, collapse = ''))



mdict$ch <- apply(expand.grid('1' = nucl, '2' = nucl, '3' = nucl,
                               '4' = 'C',  '5' = c('A', 'T', 'C'),
                               '6' = nucl, '7' = nucl, '8' = nucl),
                   1,
                   function(x) paste(x, collapse = ''))

for (item in names(mdict)) {
    fd <- mdict[[item]]
    fd <- sort(as.character(fd))
    fd <- data.frame(motif = fd, count = 0)
    rownames(fd) <- fd$motif
    mdict[[item]] <- fd
    rm(fd)
}

## filling with zeroes the motifs that are not present in some datasets

for (context in c('cg', 'ch')) {
    for (genotype in dirname(sources)) {
        for (state in c('meth', 'unmeth')) {
            curr <- d[[context]][[genotype]][[state]]

            mis <- setdiff(rownames(mdict[[context]]), rownames(curr))
            curr <- rbind(curr, mdict[[context]][mis,])
            ## lexicographical sorting
            curr <- curr[order(as.character(curr$motif)),]
            d[[context]][[genotype]][[state]] <- curr
            
            
            stopifnot(nrow(curr) == nrow(mdict[[context]]))
            rm(curr, mis)
        }
    }
}

```

Saving the data as an R object with the following structure:

- context (cg or ch)
- sample (i.e. qko)
- status (meth or unmeth)
- dataframe the columns motif, number of instances

```{r saving_r_data}

save(x = d, file = file.path(WD, 'counts_nested_list_quickchanges.RData'))
## str(d)

```

# Simplest visualization

```{r deviations_functions}


proportion <- function(x){
   rs <- sum(x);
   return(x / rs);
}


deviation_from_proportion <- function(x){
   rs <- sum(x);
   ## return((x / rs) - 0.25)
   return((x / rs) / 0.25)
}


log2ratio_proportion <- function(x){
   rs <- sum(x);
   ## return((x / rs) - 0.25)
   return(log2((x / rs) - 0.25))
}

```
## Normalized by covered/unmethylated

Getting the nucleotide weights for all the methylated and unmethylated motifs as when building a PWM. Getting the log2 enrichment of the meth/unmeth ratio.


```{r deviations_compared_to_unmeth, results = "asis"}


for (context in names(d)) {
    melted <- list()
    for (genotype in names(d[[context]])) {
        print(genotype)
        
        selected <- d[[context]][[genotype]]$meth

        for_pwm <- data.frame(A = rep(NA, nchar(selected$motif[1])),
                              C = NA,
                              G = NA,
                              T = NA)

        for (i in 1:nchar(selected$motif[1])) {
            id <- paste0('d',i)
            selected[, id] <- substr(selected$motif, i, i)
            for (nt in c('A', 'T', 'C', 'G')) {
                tsum <- 0
                tsum <- sum(selected[selected[,id] == nt,'count'])
                for_pwm[i, nt] <- tsum 
                
            }
        }
        pwm <- makePWM(apply(for_pwm, 1, proportion))

        print(sprintf('meth proportions sample %s context %s', genotype, context))
        print(kable(pwm@pwm))
        cat("\n")
        
        pwm_meth <- pwm

        ## same for unmeth

        selected <- d[[context]][[genotype]]$unmeth

        for_pwm <- data.frame(A = rep(NA, nchar(selected$motif[1])),
                              C = NA,
                              G = NA,
                              T = NA)

        for (i in 1:nchar(selected$motif[1])) {
            id <- paste0('d',i)
            selected[, id] <- substr(selected$motif, i, i)
            for (nt in c('A', 'T', 'C', 'G')) {
                tsum <- 0
                tsum <- sum(selected[selected[,id] == nt,'count'])
                for_pwm[i, nt] <- tsum 
                
            }
        }
        pwm <- makePWM(apply(for_pwm, 1, proportion))
        print(sprintf('unmeth proportions sample %s context %s', genotype, context))
        
        print(kable(pwm@pwm))
        cat("\n")
        
        pwm_unmeth <- pwm

        print(sprintf('log2 (meth/unmeth) sample %s context %s', genotype, context))
        deviations <- log2(pwm_meth@pwm /pwm_unmeth@pwm)

        print(kable(deviations))
        cat("\n")
        
        data.m <- melt(deviations)
        
        data.m$genotype <- genotype
        melted[[genotype]] <- data.m

        
    }
    melted2 <- do.call(rbind.data.frame, melted)
    colnames(melted2) <- c('nucleotide', 'position', 'deviation', 'genotype')
   
    print(ggplot(melted2, aes(position, deviation, fill=as.factor(genotype)))+
          geom_bar(position="dodge",stat="identity") +
          scale_x_continuous(breaks = 1:8) +
          facet_wrap(~nucleotide,nrow=4) + xlab("position") + ylab('log2fc meth/unmeth'))

    print(ggplot(melted2, aes(position, deviation, fill=as.factor(nucleotide)))+
          geom_bar(position="dodge",stat="identity") +
          scale_x_continuous(breaks = 1:8) +
          facet_wrap(~genotype,nrow=4) + xlab("position") + ylab('log2fc meth/unmeth'))
    
}


```


# Intermediate data download

* [Download whole RData](`r sprintf('counts_nested_list_quickchanges.RData')`)
* [Download normalized CG data](`r sprintf('cg_normalized_data_quickchanges.csv') `)
* [Download normalized CH data](`r sprintf('ch_normalized_data_quickchanges.csv') `)

# Timestamp

```{r sessionInfo}
date()
sessionInfo()
devtools::session_info()

```
