---
title: "CG and CH meth vs unmeth contexts"
author: "Izaskun Mallona"
output:
  html_document:
    keep_md: true
    toc: true
    toc_float: true
    toc_depth: 4
params:
  seed: 11

---

Per Baubec request (October 17th 2019)

Largely based upon `cph_quickchanges_simpler.Rmd`

Requires `extract_motifs_frequency_from_bam_binary.sh` to be run first.

Should contain the data already here plus the following

```
annot <- data.frame(seq = c('20181207.A-WGBS_IL12',
                            '20181207.A-WGBS_IL14',
                            '20181207.A-WGBS_IL18',
                            '20181207.A-WGBS_IL19'),
                    genotype = c('WT_DNMT3A2_in_TKO',
                                 'QC_DNMT3A2_in_TKO',
                                 'WT_DNMT3A2_in_TKO',
                                 'QC_DNMT3A2_in_TKO'))

```

```{r settings, include=FALSE}

library(knitr)
## library(devtools)
library(reshape2)
## library(lattice)
library(pheatmap)
library(seqLogo)
## library(margins)
## library(data.table)
library(ggplot2)
library(ggfortify)
library(ggseqlogo)
library(gridExtra)
library(DT)

TASK <- "cg_context"
HOME <- '/home/imallona'
SUBTASK <- 'nov_2019'

## WD <-  file.path(HOME, 'mnt', 'nfs', TASK)
WD <-  file.path('/home', 'Shared_s3it', 'imallona', TASK, SUBTASK)

## DATA <- file.path(HOME, 'data')

## MIN_DEPTH <- 10
## NFS <- file.path(HOME, 'mnt', 'nfs')

beta2m <- function(beta) {
    m <- log2(beta/(1 - beta))
    return(m)
}

m2beta <- function(m) {
    beta <- 2^m/(2^m + 1)
    return(beta)
}

opts_chunk$set(fig.width = 12,
               fig.height = 12,
               fig.path = file.path(WD, TASK, 'RmdFigs'),
               output.dir = WD,
               root.dir = WD,
               cache = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               message = TRUE)



```

# Aim

- Hypothesis: DNMTs (DNA meth transferases) show sequence specificity
-   Datasets: WGBS in three mouse cells lacking DNA meth (called
    `triple KOs or TKOs`)
-   They have a different DNMT (DNA meth transferases) addback each

# Design

```{r design, include = TRUE, cache = TRUE, fig.width = 20}
knitr::include_graphics("/home/imallona/src/cg_context/stranded_dnameth.png")

```

-   Flow
    -   Call DNA methylation for every CpG in a strand-specific manner
    -   Collapse counts (methylated/unmethylated) for each context 6-mer
    `NNNCGNNN`
    - Get nucleotide proportions for each position (i.e. 1 to 8) for methylated and unmethylated motifs separately
    - Get a log2 ratio

  
# Datasets load



```{r}
## targets <- c('tko+d3a1_merged',
##              'tko+d3a2_merged',
##              'tko+d3b1_merged',
##              'qko+d3b1_merged',
##              '20181207.A-WGBS_IL12',
##              '20181207.A-WGBS_IL14',
##              '20181207.A-WGBS_IL18',
##              '20181207.A-WGBS_IL19',
##              '20151223.B-MmES_TKOD3A1c1-3_R',
##              '20180913.A-WGBS_3',
##              '20180913.A-WGBS_4')

sources <- grep('cg_context|neuro', list.files(WD, recursive = TRUE, pattern = 'motif_counts_*'),
                invert = TRUE, value = TRUE)

## sources_targets <- unique (grep(paste(targets, collapse="|"), 
##                                 sources, value = TRUE))

```

```{r data_import, include = TRUE, echo = FALSE, cache = TRUE, cache.lazy = FALSE}

## meth and unmeth separately
d <- list(cg = list(),
          ch = list())

for (context in c('cg', 'ch')) {
    for (genotype in dirname(sources)) {
        for (state in c('meth', 'unmeth')) {
            fn <- list.files(file.path(WD, genotype),
                             pattern = sprintf('.*motif_counts_%s_%s.*', context, state))
                
            fd <- read.table(
                file.path(WD, genotype, fn),
                col.names = c('count', 'motif'),
                stringsAsFactors = FALSE)

            ## removing Ns and short motifs
            rownames(fd) <- fd$motif
            fd <- fd[grep('N', fd$motif, invert = TRUE),]
            fd <- fd[nchar(fd$motif) == 8,]
            fd <- fd[sort(fd$motif),]
            
            d[[context]][[genotype]][[state]] <- fd
            rm(fd)
        }
    }
}

```


``` {r preproc_countings}
## dictionary of the whole set of motifs
nucl <- c('T', 'C', 'G', 'A')

mdict <- list('cg' = NULL, 'ch' = NULL)
mdict$cg <- apply(expand.grid('1' = nucl, '2' = nucl, '3' = nucl,
                               '4' = 'C',  '5' = 'G',
                               '6' = nucl, '7' = nucl, '8' = nucl ),
                   1,
                   function(x) paste(x, collapse = ''))



mdict$ch <- apply(expand.grid('1' = nucl, '2' = nucl, '3' = nucl,
                               '4' = 'C',  '5' = c('A', 'T', 'C'),
                               '6' = nucl, '7' = nucl, '8' = nucl),
                   1,
                   function(x) paste(x, collapse = ''))

for (item in names(mdict)) {
    fd <- mdict[[item]]
    fd <- sort(as.character(fd))
    fd <- data.frame(motif = fd, count = 0)
    rownames(fd) <- fd$motif
    mdict[[item]] <- fd
    rm(fd)
}

## filling with zeroes the motifs that are not present in some datasets

for (context in c('cg', 'ch')) {
    for (genotype in dirname(sources)) {
        for (state in c('meth', 'unmeth')) {
            curr <- d[[context]][[genotype]][[state]]

            mis <- setdiff(rownames(mdict[[context]]), rownames(curr))
            curr <- rbind(curr, mdict[[context]][mis,])
            ## lexicographical sorting
            curr <- curr[order(as.character(curr$motif)),]
            d[[context]][[genotype]][[state]] <- curr
            
            
            stopifnot(nrow(curr) == nrow(mdict[[context]]))
            rm(curr, mis)
        }
    }
}

```

Saving the data as an R object with the following structure:

- context (cg or ch)
- sample (i.e. qko)
- status (meth or unmeth)
- dataframe the columns motif, number of instances

[/home/Shared_s3it/imallona/cg_context/nov_2019/counts_nested_list.RData](Download RData).

```{r saving_r_data}

save(x = d, file = file.path(WD, 'counts_nested_list.RData'))
## str(d)

```

# Simplest visualization

```{r deviations_functions}


proportion <- function(x){
   rs <- sum(x);
   return(x / rs);
}


deviation_from_proportion <- function(x){
   rs <- sum(x);
   ## return((x / rs) - 0.25)
   return((x / rs) / 0.25)
}


log2ratio_proportion <- function(x){
   rs <- sum(x);
   ## return((x / rs) - 0.25)
   return(log2((x / rs) - 0.25))
}

```
## Normalized by covered/unmethylated

Getting the nucleotide weights for all the methylated and unmethylated motifs as when building a PWM. Getting the log2 enrichment of the meth/unmeth ratio.


```{r deviations_compared_to_unmeth, results = "asis"}


for (context in names(d)) {
    melted <- list()
    for (genotype in names(d[[context]])) {
        print(genotype)
        
        selected <- d[[context]][[genotype]]$meth

        for_pwm <- data.frame(A = rep(NA, nchar(selected$motif[1])),
                              C = NA,
                              G = NA,
                              T = NA)

        for (i in 1:nchar(selected$motif[1])) {
            id <- paste0('d',i)
            selected[, id] <- substr(selected$motif, i, i)
            for (nt in c('A', 'T', 'C', 'G')) {
                tsum <- 0
                tsum <- sum(selected[selected[,id] == nt,'count'])
                for_pwm[i, nt] <- tsum 
                
            }
        }
        pwm <- makePWM(apply(for_pwm, 1, proportion))

        print(sprintf('meth proportions sample %s context %s', genotype, context))
        print(kable(pwm@pwm))
        cat("\n")
        
        pwm_meth <- pwm

        ## same for unmeth

        selected <- d[[context]][[genotype]]$unmeth

        for_pwm <- data.frame(A = rep(NA, nchar(selected$motif[1])),
                              C = NA,
                              G = NA,
                              T = NA)

        for (i in 1:nchar(selected$motif[1])) {
            id <- paste0('d',i)
            selected[, id] <- substr(selected$motif, i, i)
            for (nt in c('A', 'T', 'C', 'G')) {
                tsum <- 0
                tsum <- sum(selected[selected[,id] == nt,'count'])
                for_pwm[i, nt] <- tsum 
                
            }
        }
        pwm <- makePWM(apply(for_pwm, 1, proportion))
        print(sprintf('unmeth proportions sample %s context %s', genotype, context))
        
        print(kable(pwm@pwm))
        cat("\n")
        
        pwm_unmeth <- pwm

        print(sprintf('log2 (meth/unmeth) sample %s context %s', genotype, context))
        deviations <- log2(pwm_meth@pwm /pwm_unmeth@pwm)

        print(kable(deviations))
        cat("\n")
        
        data.m <- melt(deviations)
        
        data.m$genotype <- genotype
        melted[[genotype]] <- data.m

        
    }
    melted2 <- do.call(rbind.data.frame, melted)
    colnames(melted2) <- c('nucleotide', 'position', 'deviation', 'genotype')
   
    print(ggplot(melted2, aes(position, deviation, fill=as.factor(genotype)))+
          geom_bar(position="dodge",stat="identity") +
          scale_x_continuous(breaks = 1:8) +
          facet_wrap(~nucleotide,nrow=4) + xlab("position") + ylab('log2fc meth/unmeth'))

    print(ggplot(melted2, aes(position, deviation, fill=as.factor(nucleotide)))+
          geom_bar(position="dodge",stat="identity") +
          scale_x_continuous(breaks = 1:8) +
          facet_wrap(~genotype,nrow=4) + xlab("position") + ylab('log2fc meth/unmeth'))

    ## DT::datatable(melted2 %>% as.data.frame() %>% 
    ##           dplyr::mutate_if(is.numeric, funs(round(., 2))), 
    ##           extensions = c("Buttons", "FixedColumns"),
    ##           rownames = FALSE, 
    ##           options = list(dom = "Bfrtip",
    ##                          scrollX = TRUE, 
    ##                          fixedColumns = list(leftColumns = 1),
    ##                          buttons = c("csv", "excel")))
    
}


```


# Timestamp

```{r sessionInfo}
date()
sessionInfo()
devtools::session_info()

```
