---
title: "03_stats_assessment CG and CH seq preferences (no coverage filtering)"
author: "Izaskun Mallona"
output:
  html_document:
    keep_md: true
    toc: true
    toc_float: true
    toc_depth: 4
params:
  seed: 11

---

Requires `02_motif_extract_run_nov_2019_no_coverage_filtering_postproc.Rmd` to be run first.

Motifs are classified as methylated or unmethylated and counted, without aggregating average methylation values per sample to avoid batch/genotype effects.

Started 27 Feb 2020

# Config


```{r settings, include=FALSE}
library(knitr)
library(ggplot2)
library(ggExtra)
library(DT)
library(doParallel)

```

```{r}
opts_chunk$set(fig.width = 5,
               fig.height = 5,
               dpi = 300,
               cache = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               message = TRUE)



```


```{r}
options(bitmapType="cairo")

```

```{r}
NUM_PERM = 1000
NTHREADS <- 30

```
# Data retrieval

The object was generated by  `02_motif_extract_run_nov_2019_no_coverage_filtering_postproc.Rmd`

It is tracked at `./data/counts_nested_list.RData` as well

```{r}
con <- url('http://imlstaupo.uzh.ch/imallona/baubec/cg_context/counts_nested_list.RData')
load(con)
close(con)
```

## Permutation analysis on `tko+d3a2_merged`

To check whether the proportion of methylated sites versus the unmethylated sites (or versus the coverage), we need to run a test that maintains the average DNA methylation of that sample.

We do this in two ways.

- (permute motifs strategy) First, permuting the motifs a certain number of times.
- (permute 2d strategy) Or permuting both the methylated and the unmethylated counts a certain number of times.

The first approach is the soundest to detect motifs that deviate from the observations.

```{r, d3acg}

## keeping_meth_and_unmeth_blocks
permute_motifs <- function(fd, n = 1000, nthreads = 2) {

    res <- list()

    cl <- makeCluster(nthreads)

    foreach(i = 1:n) %do% {
        idx <- sample(x = rownames(fd), size = nrow(fd), replace = FALSE)
        res[[i]] <- data.frame(motif = fd$motif,
                               meth = fd[idx, 'meth'],
                               unmeth = fd[idx, 'unmeth'])

        rownames(res[[i]]) <- res[[i]]$motif

        res[[i]]$score <- res[[i]]$meth / (res[[i]]$meth  + res[[i]]$unmeth)
    }

    stopCluster(cl)
    
    return(res)
}


permute_2d <- function(fd, n = 1000, nthreads = 2) {

    res <- list()

    cl <- makeCluster(nthreads)
    
    foreach(i = 1:n) %do% {
        idx <- sample(x = rownames(fd), size = nrow(fd), replace = FALSE)
        idx2 <- sample(x = rownames(fd), size = nrow(fd), replace = FALSE)
        res[[i]] <- data.frame(motif = fd$motif,
                               meth = fd[idx, 'meth'],
                               unmeth = fd[idx2, 'unmeth'])

        rownames(res[[i]]) <- res[[i]]$motif

        res[[i]]$score <- res[[i]]$meth / (res[[i]]$meth  + res[[i]]$unmeth)
    }

    stopCluster(cl)
    
    return(res)
}

## note that permutation p-values should never be 0;
## we add `1` to the one-tailed test to include the observation on top to the
##   number of permutations
run_permutation_test <- function(fd, permutation_type, n = 1000, adjust.FDR = TRUE,
                                 nthreads = 2) {
    if (permutation_type == 'motifs') {
        res <- permute_motifs(fd = fd, n = n, nthreads = nthreads)
    }
    else if (permutation_type == '2d') {
        res <- permute_2d(fd, n = n, nthreads = nthreads)
    }
    else {
        stop('wrong permutation type')
    }

    fd$pval <- NA
    fd$qval <- NA ## Benjamini Hochberg corrected p-val

    cl <- makeCluster(nthreads)
    
    foreach (i = 1:nrow(fd)) %do% {
        motif <- rownames(fd)[i]
        fd[motif, 'pval'] <- (sum(
            sapply(res,
                   function(x) fd[motif,]$score > x[motif,]$score)) + 1) / (n + 1)
        if(fd[motif, 'pval'] > 1) stop()
       
    }

    stopCluster(cl)

    if (adjust.FDR) {
        fd$qval <- p.adjust(fd$pval, method = 'BH')
    }

    return(fd)     
}


```

# Permutation analysis on `tko+d3a2_merged`

## Permuted motifs

Permuting motifs in DNMT D3A2 

```{r}
stopifnot(all(d$cg$`tko+d3a2_merged`$meth$motif == d$cg$`tko+d3a2_merged`$unmeth$motif))

a <- data.frame(motif = d$cg$`tko+d3a2_merged`$meth$motif,
                meth = d$cg$`tko+d3a2_merged`$meth$count,
                unmeth = d$cg$`tko+d3a2_merged`$unmeth$count)

rownames(a) <- a$motif
a$score <- a$meth / (a$meth + a$unmeth)
head(a$score)

a_motif_perm <- run_permutation_test(fd = a,
                                     n = NUM_PERM,
                                     permutation_type = 'motifs',
                                     adjust.FDR = TRUE,
                                     nthreads = NTHREADS)


summary(a_motif_perm)
table(a_motif_perm$pval < 0.05)
table(a_motif_perm$qval < 0.05)
```

```{r}
DT::datatable(a_motif_perm %>% as.data.frame() %>% 
              dplyr::mutate_if(is.numeric, funs(round(., 2))), 
              extensions = c("Buttons", "FixedColumns"),
              rownames = FALSE, 
              options = list(dom = "Bfrtip",
                             scrollX = TRUE, 
                             fixedColumns = list(leftColumns = 1),
                             buttons = c("csv", "excel")))
```    

Plots with marginals


```{r}
p <- ggplot(a_motif_perm, aes(y = score, x = pval)) +
    geom_point() +
    theme_classic(base_size = 20) +
    ylab('methylation score') +
    xlab('p-value (permuted motifs)') +
    ggtitle('DNMT3A2')

ggMarginal(p, type = "histogram", bins = 10, colour = 'black')#, groupColour = TRUE)

```


```{r}
p <- ggplot(a_motif_perm, aes(x=qval, y=score)) +
    geom_point() +
    theme_classic(base_size = 20) +
    ylab('methylation score') +
    xlab ('FDR (permuted motifs)') +
    ggtitle('DNMT3A2')

ggMarginal(p, type = "histogram", bins = 10, colour = 'black')

```


## Permuting both meth and unmeth counts

Permuting meth and unmeth counts in DNMT D3A2

```{r}

a_all_perm <- run_permutation_test(fd = a,
                                   n = NUM_PERM,
                                   permutation_type = '2d',
                                   adjust.FDR = TRUE,
                                   nthreads = NTHREADS)

summary(a_all_perm)
table(a_all_perm$pval < 0.05)
table(a_all_perm$qval < 0.05)

```

```{r}
DT::datatable(a_all_perm %>% as.data.frame() %>% 
              dplyr::mutate_if(is.numeric, funs(round(., 2))), 
              extensions = c("Buttons", "FixedColumns"),
              rownames = FALSE, 
              options = list(dom = "Bfrtip",
                             scrollX = TRUE, 
                             fixedColumns = list(leftColumns = 1),
                             buttons = c("csv", "excel")))
```


```{r}
p <- ggplot(a_all_perm, aes(y = score, x = pval)) +
    geom_point() +
    theme_classic(base_size = 20) +
    ylab('methylation score') +
    xlab('p-value (permuted counts)') +
    ggtitle('DNMT3A2')

ggMarginal(p, type = "histogram", bins = 10, colour = 'black')#, groupColour = TRUE)

```


```{r}
p <- ggplot(a_all_perm, aes(x=qval, y=score)) +
    geom_point() +
    theme_classic(base_size = 20) +
    ylab('methylation score') +
    xlab ('FDR (permuted counts)') +
    ggtitle('DNMT3A2')

ggMarginal(p, type = "histogram", bins = 10, colour = 'black')

```



## Comparison of both strategies

Comparing both permutation strategies.

```{r, fig.width = 8, fig.height = 4}

par(mfrow = c(3,1))
hist(a_motif_perm$pval)
hist(a_all_perm$pval)

plot(a_motif_perm$pval, a_all_perm$pval)

```




## Permutation analysis on `tko+d3b1_merged`

Data reshaping for tko+d3b1


```{r}
b <- data.frame(motif = d$cg$`tko+d3b1_merged`$meth$motif,
                meth = d$cg$`tko+d3b1_merged`$meth$count,
                unmeth = d$cg$`tko+d3b1_merged`$unmeth$count)

rownames(b) <- b$motif
b$score <- b$meth / (b$meth + b$unmeth)
head(b$score)

```

## Permuting motifs

Permuting motifs in tko+d3b1

```{r}
b_motif_perm <- run_permutation_test(fd = b,
                                     n = NUM_PERM,
                                     permutation_type = 'motifs',
                                     adjust.FDR = TRUE,
                                     nthreads = NTHREADS)


summary(b_motif_perm)
table(b_motif_perm$pval < 0.05)
table(b_motif_perm$qval < 0.05)
```


```{r}
DT::datatable(b_motif_perm %>% as.data.frame() %>% 
              dplyr::mutate_if(is.numeric, funs(round(., 2))), 
              extensions = c("Buttons", "FixedColumns"),
              rownames = FALSE, 
              options = list(dom = "Bfrtip",
                             scrollX = TRUE, 
                             fixedColumns = list(leftColumns = 1),
                             buttons = c("csv", "excel")))
```

```{r}
p <- ggplot(b_motif_perm, aes(y = score, x = pval)) +
    geom_point() +
    theme_classic(base_size = 20) +
    ylab('methylation score') +
    xlab('p-value (permuted motifs)') +
    ggtitle('DNMT3B1')

ggMarginal(p, type = "histogram", bins = 10, colour = 'black')#, groupColour = TRUE)

```


```{r}
p <- ggplot(b_motif_perm, aes(x=qval, y=score)) +
    geom_point() +
    theme_classic(base_size = 20) +
    ylab('methylation score') +
    xlab ('FDR (permuted motifs)') +
    ggtitle('DNMT3B1')

ggMarginal(p, type = "histogram", bins = 10, colour = 'black')

```


## Permuting both meth and unmeth counts

Permuting meth and unmeth counts for tko+d3b1

```{r}

b_all_perm <- run_permutation_test(fd = b,
                                   n = NUM_PERM,
                                   permutation_type = '2d',
                                   adjust.FDR = TRUE,
                                   nthreads = NTHREADS)

summary(b_all_perm)
table(b_all_perm$pval < 0.05)
table(b_all_perm$qval < 0.05)
```



```{r}
DT::datatable(b_all_perm %>% as.data.frame() %>% 
              dplyr::mutate_if(is.numeric, funs(round(., 2))), 
              extensions = c("Buttons", "FixedColumns"),
              rownames = FALSE, 
              options = list(dom = "Bfrtip",
                             scrollX = TRUE, 
                             fixedColumns = list(leftColumns = 1),
                             buttons = c("csv", "excel")))
```

```{r}
p <- ggplot(b_all_perm, aes(y = score, x = pval)) +
    geom_point() +
    theme_classic(base_size = 20) +
    ylab('methylation score') +
    xlab('p-value (permuted counts)') +
    ggtitle('DNMT3A2')

ggMarginal(p, type = "histogram", bins = 10, colour = 'black')#, groupColour = TRUE)

```


```{r}
p <- ggplot(b_all_perm, aes(x=qval, y=score)) +
    geom_point() +
    theme_classic(base_size = 20) +
    ylab('methylation score') +
    xlab ('FDR (permuted counts)') +
    ggtitle('DNMT3A2')

ggMarginal(p, type = "histogram", bins = 10, colour = 'black')

```



## Comparing both permutation strategies.

```{r}
par(mfrow = c(2,2), pty = 's')
hist(b_motif_perm$pval)
hist(b_all_perm$pval)

plot(b_motif_perm$pval, b_all_perm$pval)

plot(b_motif_perm$pval[b_motif_perm$pval < 0.05 | b_all_perm$pval < 0.05 ],
     b_all_perm$pval[b_motif_perm$pval < 0.05 | b_all_perm$pval < 0.05 ])
```

# Comparing DNMT3A and DNMT3B meth preferences according to the permutation outputs


```{r}
p <- ggplot(data = data.frame(pval_a = a_motif_perm$pval,
                              pval_b = b_motif_perm$pval),
                aes(x=pval_a, y = pval_b)) +
    geom_point() +
    theme_classic(base_size = 20) +
    ylab('pval D3A2') +
    xlab ('pval D3B1') + ggtitle('permuted motifs')


ggMarginal(p, type = "histogram", bins = 10, colour = 'black')
    

```


```{r}
p <- ggplot(data = data.frame(qval_a = a_motif_perm$qval,
                              qval_b = b_motif_perm$qval),
                aes(x=qval_a, y = qval_b)) +
    geom_point() +
    theme_classic(base_size = 20) +
    ylab('FDR D3A2') +
    xlab ('FDR D3B1') + ggtitle('permuted motifs')


ggMarginal(p, type = "histogram", bins = 10, colour = 'black')
    

```


```{r}
p <- ggplot(data = data.frame(pval_a = a_all_perm$pval,
                              pval_b = b_all_perm$pval),
                aes(x=pval_a, y = pval_b)) +
    geom_point() +
    theme_classic(base_size = 20) +
    ylab('pval D3A2') +
    xlab ('pval D3B1') + ggtitle('permuted counts')


ggMarginal(p, type = "histogram", bins = 10, colour = 'black')
    

```


```{r}
p <- ggplot(data = data.frame(qval_a = a_all_perm$qval,
                              qval_b = b_all_perm$qval),
                aes(x=qval_a, y = qval_b)) +
    geom_point() +
    theme_classic(base_size = 20) +
    ylab('FDR D3A2') +
    xlab ('FDR D3B1') + ggtitle('permuted counts')


ggMarginal(p, type = "histogram", bins = 10, colour = 'black')
    

```



# DNMT3A vs DNMT3B by proportion tests

We'll check the methylation success per motif according to the enzyme added back. We'll directly compare DNMT3A and DNMT3B.

Because of the upper plot (section `Comparing DNMT3A and DNMT3B meth preferences according to the permutation outputs`) it's expected that success rate (proportion of methylated vs unmethylated) for most of the motifs will be different. The H0 (null of proportions being the same in DNMT3A and DNMT3B) will most probably be rejected for most cases.


```{r}
stopifnot(all(d$cg$`tko+da3a2_merged`$meth$motif == d$cg$`tko+d3b1_merged`$unmeth$motif))
```

Assuming independences and H0: methylation proportion in DNMT3A and DNMT3b are the same, iterating all motifs as follows?

```{r}

prop_test_to_iterate <- function(a, b, motif) {
    curr <- as.data.frame(matrix(data = NA, nrow = 2, ncol = 2))
    
    curr[1,] <- as.numeric(a[motif, c('meth', 'unmeth')])
    curr[2,] <- as.numeric(b[motif, c('meth', 'unmeth')])
    
    colnames(curr) <- c('meth', 'unmeth')
    rownames(curr) <- sprintf('%s_%s', motif, c('a', 'b'))
    
    return(list(table = curr,
                test = prop.test(x = curr$meth, n = curr$meth + curr$unmeth, correct = TRUE)))
}

## prop_test_to_iterate(a = a, b = b, motif = 'AAACGCGC')

```

```{r}

ptests_list <- list()

for (motif in a$motif) {
    ptests_list[[motif]] <- prop_test_to_iterate(a = a, b = b, motif = motif)
}


ptests <- data.frame(motif = names(sapply(ptests_list, function(x) return(x$test$p.value))),
                  pval = as.numeric(sapply(ptests_list, function(x) return(x$test$p.value))))

rownames(ptests) <- ptests$motif

ptests$qval <- p.adjust(ptests$pval, method = 'BH')

```

Almost all proportion tests are different

```{r}
table(ptests$qval < 0.05)

```



```{r}
DT::datatable(ptests %>% as.data.frame() %>% 
              dplyr::mutate_if(is.numeric, funs(round(., 2))), 
              extensions = c("Buttons", "FixedColumns"),
              rownames = FALSE, 
              options = list(dom = "Bfrtip",
                             scrollX = TRUE, 
                             fixedColumns = list(leftColumns = 1),
                             buttons = c("csv", "excel")))
```

```{r, fig.height = 4, fig.width = 8}
par(mfrow = c(1,2), pty = 's')
hist(ptests$pval)
hist(ptests$qval)

```

## Plot z-scores and plot the score difference DNMTA- DNMTB 



```{r zscores}

## modified from gCMAP::zScores
zscore_from_pval <- function (pval, direction = NULL, tails = 1, limit = .Machine$double.xmin){ 
    if (!is.null(limit)) {
        pval[which(pval < limit)] <- limit
    }
    if (tails == 2) {
        z <- qnorm(pval/2, lower.tail = FALSE)
    }
    else if (tails == 1) {
        z <- qnorm(pval, lower.tail = FALSE)
    }
    else {
        stop("Parameter 'tails' must be set to either 1 or 2.")
    }

    ## direction of change
    if (!is.null(direction)) {
        z <- z * sign(direction)
    }
    return(z)
}


```


```{r, eval = TRUE}
ptests$delta <- a[rownames(ptests), 'score'] - b[rownames(ptests), 'score']
ptests$score_dnmt3a <-  a[rownames(ptests), 'score']
ptests$score_in_dnmt3b <- b[rownames(ptests), 'score']
ptests$zscore <- zscore_from_pval(ptests$pval, direction = ptests$delta, tails = 2)

summary(ptests)

```

```{r, fig.width = 8, fig.height = 8}
## plot(ptests)
par(mfrow = c(3,1), pty = 's')

plot(ptests$delta, ptests$zscore, pch = 19)
plot(ptests$score_dnmt3a, ptests$zscore, pch = 19)
plot(ptests$score_in_dnmt3b, ptests$zscore, pch = 19)

```

Sort the motifs by their zscore and then the delta of the scores, and plot the trends


```{r}
ptests <- ptests[order(abs(ptests$zscore), ptests$delta,
                               decreasing = TRUE),]
ptests$zscore_rank <- 1:nrow(ptests)
```


```{r, fig.width = 8, fig.weight = 8}

par(mfrow = c(2,2), pty = 's')
plot(y = abs(ptests$zscore), x = ptests$zscore_rank)
plot(y = abs(ptests$delta), x = ptests$zscore_rank)
plot(y = abs(ptests$score_in_dnmt3b), x = ptests$zscore_rank)
plot(y = abs(ptests$score_dnmt3a), x = ptests$zscore_rank)

```


Compare ptests and permutation results?

# Timestamp

```{r sessionInfo, cache = 0}
date()
sessionInfo()
devtools::session_info()

```
