---
title: "05_ma_plots CG and CH seq preferences (no coverage filtering)"
author: "Izaskun Mallona"
output:
  html_document:
    keep_md: true
    toc: true
    toc_float: true
    toc_depth: 4
params:
  seed: 11

---

Requires `02_motif_extract_run_nov_2019_no_coverage_filtering_postproc.Rmd` to be run first.

Motifs are classified as methylated or unmethylated and counted, without aggregating average methylation values per sample to avoid batch/genotype effects.

Started 11 March 2020

# Config

<details><summary>Expand</summary>

Libraries, knitr options, permutation numbers, threads.

```{r settings, include=FALSE}
library(knitr)
library(Cairo)
library(ggplot2)
library(DT)
library(ggExtra)
library(dplyr)
library(gridExtra)

```

```{r}
opts_chunk$set(fig.width = 5,
               fig.height = 5,
               dpi = 200,
               cache = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               message = TRUE)



```


```{r}
options(bitmapType="cairo")

```

```{r, functions}

ac <- function(col, alpha=1){

  apply(sapply(col, col2rgb)/255, 2, 
                     function(x) 
                       rgb(x[1], x[2], x[3], alpha=alpha))  
}

get_palette <- function(alpha = alpha) {
    data.frame(
        nucleotide = c('G', 'C', 'A', 'T'),
        color = ac(c('#F6A318', '#2B4C9B',  '#68B42E', '#E52521'),
                   alpha),
        stringsAsFactors = FALSE)
}

color_by_nucleotide_and_position <- function(motifs, start, alpha){
    
    palette <- get_palette(alpha = alpha)
    
    tmp <- data.frame(motif = as.character(motifs),
                      nucleotide = substr(as.character(motifs), start, start ))


    tmp2 <- merge(tmp, palette, by.x = 'nucleotide', by.y = 'nucleotide', all.x = TRUE)
    rownames(tmp2) <- tmp2$motif

    tmp3 <- ac(tmp2[as.character(motifs), 'color'], alpha = alpha)
    names(tmp3) <- as.character(tmp2[as.character(motifs), 'nucleotide'])
    return(tmp3)  
}

color_by_nucleotide_and_position(c('ACG', 'CGT'), 1, 1)

```

</details>

# Data retrieval

The object was generated by  `02_motif_extract_run_nov_2019_no_coverage_filtering_postproc.Rmd`

```{r}
con <- url('http://imlstaupo.uzh.ch/imallona/baubec/cg_context/counts_nested_list.RData')
load(con)
close(con)
```

## `tko+d3a2_merged` CG

```{r}
stopifnot(all(d$cg$`tko+d3a2_merged`$meth$motif == d$cg$`tko+d3a2_merged`$unmeth$motif))

a <- data.frame(motif = d$cg$`tko+d3a2_merged`$meth$motif,
                meth = d$cg$`tko+d3a2_merged`$meth$count,
                unmeth = d$cg$`tko+d3a2_merged`$unmeth$count)

rownames(a) <- a$motif
a$score <- a$meth / (a$meth + a$unmeth)
a$cov <- a$meth + a$unmeth

## shuffling for plotting later
set.seed(2)
a <- a[sample(rownames(a), nrow(a), replace = FALSE),]

```


```{r}
DT::datatable(a %>% as.data.frame() %>% 
              dplyr::mutate_if(is.numeric, funs(round(., 2))), 
              extensions = c("Buttons", "FixedColumns"),
              rownames = FALSE, 
              options = list(dom = "Bfrtip",
                             scrollX = TRUE, 
                             fixedColumns = list(leftColumns = 1),
                             buttons = c("csv", "excel")))
```

## `tko+d3b1_merged` CG


```{r}
b <- data.frame(motif = d$cg$`tko+d3b1_merged`$meth$motif,
                meth = d$cg$`tko+d3b1_merged`$meth$count,
                unmeth = d$cg$`tko+d3b1_merged`$unmeth$count)

rownames(b) <- b$motif
b$score <- b$meth / (b$meth + b$unmeth)
b$cov <- b$meth + b$unmeth

## shuffling for plotting later
set.seed(2)
b <- b[sample(rownames(b), nrow(b), replace = FALSE),]
```

```{r}
DT::datatable(b %>% as.data.frame() %>% 
              dplyr::mutate_if(is.numeric, funs(round(., 2))), 
              extensions = c("Buttons", "FixedColumns"),
              rownames = FALSE, 
              options = list(dom = "Bfrtip",
                             scrollX = TRUE, 
                             fixedColumns = list(leftColumns = 1),
                             buttons = c("csv", "excel")))
```

# MA plots


## log2 scores vs mean coverage

```{r, fig.width = 8, fig.height = 8}

stopifnot(a$motif == b$motif)

par(mfrow = c(3, 3), pty = 's')

for (i in 1:8){
    plot( y = log2(a$score/b$score),
         x = (a$cov + b$cov)/2,
         col = color_by_nucleotide_and_position(a$motif, start = i, alpha = 0.5),
         pch = 19,
         ylab = "log2 (DNMT3A score/ DNMT3B score)",
         xlab = "mean coverage",
         main = sprintf('position %s', i))
}


```


```{r, fig.width = 10, fig.height = 10}
stopifnot(a$motif == b$motif)

fd  <- data.frame(motif = a$motif,
                  meth_dnmt3a = a$meth,
                  unmeth_dnmt3a = a$unmeth,
                  cov_dnmt3a = a$cov,
                  score_dnmt3a = a$score,                  
                  meth_dnmt3b = b$meth,
                  unmeth_dnmt3b = b$unmeth,
                  cov_dnmt3b = b$cov,
                  score_dnmt3b = b$score)

p <- list()


col_scale <- scale_colour_manual(
    name = sprintf(""),
    labels = get_palette(alpha = alpha)$nucleotide,
    values = get_palette(alpha = alpha)$color,
    drop = FALSE)


for (i in 1:8) {
    alpha = 0.3
    cols <- color_by_nucleotide_and_position(fd$motif,
                                             start = i, alpha = alpha)
    
    cols[1:3]

    fd[,sprintf('nt_%i', i)] <- factor(substr(
        as.character(fd$motif), i, i),
        levels = get_palette(alpha = alpha)$nucleotide)

    str(fd)
    
    p[[i]] <- ggplot(fd, aes(y = log2(score_dnmt3a/score_dnmt3b),
                             x = (cov_dnmt3a + cov_dnmt3b)/2,
                             colour = get(sprintf('nt_%s', i)))) +
        geom_point() +
        col_scale +
        xlab('average coverage') +
        ylab('log2 (DNMT3A/DNMT3B) scores') +
        ggtitle(sprintf("Nucleotide %s", i-4)) +
        theme(legend.position = "none")

    ## + theme(legend. position = "none")
    

    p[[i]] <- ggMarginal(p[[i]], groupColour = TRUE, groupFill = TRUE)
    ## print(p[[i]])
    ## p[[i]]

}

do.call("grid.arrange", c(p, ncol=3))

         ```
## delta vs mean coverage


```{r, fig.width = 8, fig.height = 8}

stopifnot(a$motif == b$motif)

par(mfrow = c(3,3), pty = 's')

for (i in 1:8){
    plot( y = a$score - b$score,
         x = (a$cov + b$cov)/2,
         col = color_by_nucleotide_and_position(a$motif, start = i, alpha = 0.5),
         pch = 19,
         ylab = "DNMT3A score - DNMT3B score",
         xlab = "mean coverage",
         main = sprintf('position %s', i))
}
     
```

## log2 scores vs DNMT3 A score


```{r, fig.width = 8, fig.height = 8}

stopifnot(a$motif == b$motif)

par(mfrow = c(3,3), pty = 's')

for (i in 1:8){
    plot( y = log2(a$score/b$score),
         x = a$score,
         col = color_by_nucleotide_and_position(a$motif, start = i, alpha = 0.5),
         pch = 19,
         ylab = "log2(DNMT3A score / DNMT3B score)",
         xlab = "DNMT3A score",
         main = sprintf('position %s', i))
}
     
```

## delta vs DNMT3 A score


```{r, fig.width = 8, fig.height = 8}

stopifnot(a$motif == b$motif)

par(mfrow = c(3,3), pty = 's')

for (i in 1:8){
    plot( y = a$score - b$score,
         x = a$score,
         col = color_by_nucleotide_and_position(a$motif, start = i, alpha = 0.5),
         pch = 19,
         ylab = "DNMT3A score - DNMT3B score",
         xlab = "DNMT3A score",
         main = sprintf('position %s', i))
}
     
```


## log2 scores vs DNMT3 B score


```{r, fig.width = 8, fig.height = 8}

stopifnot(a$motif == b$motif)

par(mfrow = c(3,3), pty = 's')

for (i in 1:8){
    plot( y = log2(a$score/b$score),
         x = b$score,
         col = color_by_nucleotide_and_position(a$motif, start = i, alpha = 0.5),
         pch = 19,
         ylab = "log2(DNMT3A score / DNMT3B score)",
         xlab = "DNMT3B score",
         main = sprintf('position %s', i))
}
     
```

## delta vs DNMT3 B score


```{r, fig.width = 8, fig.height = 8}

stopifnot(a$motif == b$motif)

par(mfrow = c(3,3), pty = 's')

for (i in 1:8){
    plot( y = a$score - b$score,
         x = b$score,
         col = color_by_nucleotide_and_position(a$motif, start = i, alpha = 0.5),
         pch = 19,
         ylab = "DNMT3A score - DNMT3B score",
         xlab = "DNMT3B score",
         main = sprintf('position %s', i))
}
     
```

<!-- ## score quantiles DNMT3 A -->

<!-- ```{r, fig.width = 16, fig.height = 10} -->


<!-- par(mfrow = c(4,8), pty = 's') -->

<!-- for (quantile in 1:4) { -->
<!--     for (position in 1: 8) { -->
<!--         plot(y = a[a$score <]$score - b$score, -->
<!--              x = (a$cov + b$cov)/2, -->
<!--              col = color_by_nucleotide_and_position(a$motif, start = i, alpha = 0.5), -->
<!--              pch = 19, -->
<!--              ylab = "DNMT3A score - DNMT3B score", -->
<!--              xlab = "DNMT3B score", -->
<!--              main = sprintf('position %s', i)) -->

<!--         quantile(a$score[1]) -->
        
<!--     } -->
<!-- } -->

<!-- ``` -->

<!-- ## score quantiles DNMTB -->

## ggMarginal

```{r}
data(mtcars)
p <- ggplot(mtcars, aes(x = wt, y = drat, colour = factor(vs))) +
      geom_point()


ggMarginal(p, groupColour = TRUE, groupFill = TRUE)
```
# Timestamp

```{r sessionInfo, cache = 0}
date()
sessionInfo()
devtools::session_info()

```
